File: src/gestorAplicacion/gestionVuelos/Vuelo.java
// Autores: 
/*
 * 
 */

package gestorAplicacion.gestionVuelos;

import java.io.Serializable;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

import gestorAplicacion.gestionHumana.Empleado;
import gestorAplicacion.gestionHumana.Pasajero;

public class Vuelo implements Serializable {
	/**
	 * 
	 */
	private static final long serialVersionUID = 1L;
	private Avion avion;
	private List<Empleado> empleados = new ArrayList<>();
	private List<Pasajero> pasajeros = new ArrayList<>();
	private LocalDateTime fecha;
	private final static String origen = "Medellin";
	private String destino;
	private boolean enVuelo;
	private int costo;
	private String salaEmbarque;
	private double pesoActual;
	private static int globalID = 1;
	private final int id;
	private static Aeropuerto aeropuerto;

	public Vuelo(Avion avion, LocalDateTime fecha, String destino, int costo, String salaEmbarque) {
		this.avion = avion;
		this.fecha = fecha;
		this.destino = destino;
		this.costo = costo;
		this.salaEmbarque = salaEmbarque;
		aeropuerto.agregarVuelo(this);
		this.id = globalID;
		Vuelo.globalID++;
	}

	public boolean agregarPasajero(Pasajero pasajero, int nroAsiento) {

		double pesoEquipajePasajero = 0;
		for (Equipaje equipaje : pasajero.getEquipajes())
			pesoEquipajePasajero += equipaje.getPeso();
		Asiento asientoElegido = avion.getAsientos().get(nroAsiento - 1);

		if (pesoActual + pesoEquipajePasajero < avion.getPesoMaximo() && pasajeros.size() < avion.getAsientos().size()
				&& !asientoElegido.isOcupado()) {
			pesoActual += pesoEquipajePasajero;
			pasajeros.add(pasajero);
			pasajero.setAsiento(asientoElegido);
			pasajero.setVuelo(this);
			asientoElegido.setOcupado(true);
			if (asientoElegido.getClase().equals("Primera clase")) {
				aeropuerto.ingresarDinero(3 * costo);
				pasajero.setInversion(3 * costo);
			} else if (asientoElegido.getClase().equals("Ejecutiva")) {
				aeropuerto.ingresarDinero(2 * costo);
				pasajero.setInversion(2 * costo);
			} else {
				aeropuerto.ingresarDinero(costo);
				pasajero.setInversion(costo);
			}
			return true;
		} else if (pesoActual + pesoEquipajePasajero > avion.getPesoMaximo()) {
			System.out.println("\n"
					+ "No queda espacio suficiente en este vuelo para su equipaje, por favor elija otro vuelo o reduzca el peso.");
			return false;
		} else {
			System.out.println("\n" + "Ha elegido un puesto no disponible, por favor elija otro.");
			return false;
		}
	}

	@Override
	public String toString() {
		String infoVuelo = "ID vuelo: " + id + " - Fecha del vuelo: " + fecha + " - Origen: " + origen + " - Destino: "
				+ destino + " - Precio: " + costo + "\n";
		return infoVuelo;
	}

	public String tiquete(Pasajero pasajero) {
		String tique = "\n" + "Ha sido registrado exitosamente" + "\n" + "\n" + "------------------------------------\n"
				+ "          Tiquete No." + this.id + "\n" + "------------------------------------\n"
				+ "Nombre Pasajero: " + pasajero.getNombre() + "\n" + "Fecha: " + fecha + "\n" + "Vuelo: " + getId()
				+ "\n" + "Sala de embarque: " + pasajero.getVuelo().getSalaEmbarque() + "\n" + "Clase: "
				+ pasajero.getAsiento().getClase() + "\n" + "Num Silla: " + pasajero.getAsiento().getNumero() + "\n"
				+ "Origen: " + origen + "\n" + "Destino: " + getDestino() + "\n" + "Precio Total: "
				+ pasajero.getInversion() + "\n" + "------------------------------------\n";
		return tique;
	}

	public Avion getAvion() {
		return avion;
	}

	public void setAvion(Avion avion) {
		this.avion = avion;
	}

	public List<Empleado> getEmpleados() {
		return empleados;
	}

	public void setEmpleados(List<Empleado> empleados) {
		this.empleados = empleados;
	}

	public List<Pasajero> getPasajeros() {
		return pasajeros;
	}

	public void setPasajeros(List<Pasajero> pasajeros) {
		this.pasajeros = pasajeros;
	}

	public LocalDateTime getFecha() {
		return fecha;
	}

	public void setFecha(LocalDateTime fecha) {
		this.fecha = fecha;
	}

	public String getOrigen() {
		return origen;
	}

	public String getDestino() {
		return destino;
	}

	public void setDestino(String destino) {
		this.destino = destino;
	}

	public boolean isEnVuelo() {
		return enVuelo;
	}

	public void setEnVuelo(boolean enVuelo) {
		this.enVuelo = enVuelo;
	}

	public int getCosto() {
		return costo;
	}

	public void setCosto(int costo) {
		this.costo = costo;
	}

	public String getSalaEmbarque() {
		return salaEmbarque;
	}

	public void setSalaEmbarque(String salaEmbarque) {
		this.salaEmbarque = salaEmbarque;
	}

	public double getPesoActual() {
		return pesoActual;
	}

	public void setPesoActual(double pesoActual) {
		this.pesoActual = pesoActual;
	}

	public int getId() {
		return id;
	}

	public static void setAeropuerto(Aeropuerto aeropuerto) {
		Vuelo.aeropuerto = aeropuerto;
	}

	public static Vuelo encontrarVuelo(int id) {
		if (id - 1 <= aeropuerto.getVuelos().size()) {
			return aeropuerto.getVuelos().get(id - 1);
		}
		return null;
	}
}

File: python/gestorAplicacion/gestorVuelos/vuelo.py
import time

import baseDatos.deserializador as deserializador
from gestorAplicacion.gestorHumana.pasajero import Pasajero
from gestorAplicacion.gestorVuelos.aeropuerto import Aeropuerto


class Vuelo:
    _origen = "Medellín"
    _aeropuerto = None
    def __init__(self,avion,fecha,destino,costo,salaEmbarque):
        self._avion = avion
        self._empleados = []
        self._pasajeros = []
        self._fecha = fecha
        self._destino = destino
        self._envuelo = False
        self._costo = costo
        self._salaEmbarque = salaEmbarque
        self._pesoActual = 0
        self._id = 0
        self._aeropuerto.addVuelo(self)
    
    @classmethod
    def nuevoID(cls):
        if len(cls._aeropuerto.getVuelos()) != 0:
            return cls._aeropuerto.getVuelos().get(len(cls._aeropuerto.getVuelos())-1).getId() + 1

        else:
            return 1

    def tiquete(self,pasajero):
       tique = str("\n" + "Ha sido registrado exitosamente" + "\n" + "\n" + "------------------------------------\n"
				+ "             Tiquete " + "\n" + "------------------------------------\n"
				+ "Nombre Pasajero: " + pasajero.getNombre() + "\n" + "Fecha: " + self._fecha + "\n" + "Vuelo: " + self.getId()
				+ "\n" + "Sala de embarque: " + pasajero.getVuelo().getSalaEmbarque() + "\n" + "Clase: "
				+ pasajero.getAsiento().getClase() + "\n" + "Num Silla: " + pasajero.getAsiento().getNumero() + "\n"
				+ "Origen: " + self._origen + "\n" + "Destino: " + self.getDestino() + "\n" + "Precio Total: "
				+ pasajero.getInversion() + "\n" + "------------------------------------\n")
       return tique

    def getAvion(self):
        return self._avion
    
    def setAvion(self,avion):
        self._avion = avion
    
    def getEmpleados(self):
        return self._empleados

    def setEmpleados(self,empleados):
        self._empleados = empleados

    def addEmpleado(self,e):
        self._empleados.append(e)

    def getPasajeros(self):
        return self._pasajeros

    def setPasajeros(self,pasajeros):
        self._pasajeros = pasajeros
    
    def addPasajero(self,pasajero):
        self._pasajeros.append(pasajero)

    def getFecha(self):
        return self._fecha
    
    def setFecha(self,fecha):
        self._fecha = fecha

    def getDestino(self):
        return self._destino

    def setDestino(self,destino):
        self._destino = destino
    
    def isEnVuelo(self):
        return self._envuelo 
    
    def setEnVuelo(self,envuelo):
        self._envuelo = envuelo

    def getCosto(self):
        return self._costo

    def setCosto(self,costo):
        self._costo = costo

    def getSalaEmbarque(self):
        return self._salaEmbarque

    def setSalaEmbarque(self,sala):
        self._salaEmbarque = sala

    def getPesoActual(self):
        return self._pesoActual

    def setPesoActual(self,peso):
        self._pesoActual = peso

    def getId(self):
        return self._id
    @classmethod
    def setAeropuerto(cls,aeropuerto):
        cls._aeropuerto = aeropuerto

    def encontrarVuelo(self,id):
        if id - 1 <= len(Vuelo._aeropuerto.getVuelos()):
            return Vuelo._aeropuerto.getVuelos().get(id-1)

        return None